### PR TITLE
Change nef checksum to double SHA256

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/1750

### CODE REVIEW

File: src/neo/SmartContract/NefFile.cs [hunk 1]
```diff
@@ -110,20 +111,17 @@ public void Deserialize(BinaryReader reader)
...
+            wr.Flush();
 
-                Span<byte> buffer = stackalloc byte[HeaderSize - sizeof(uint)];
-                ms.Seek(0, SeekOrigin.Begin);
-                ms.Read(buffer);
+            // Read header without CRC
 
-                return BitConverter.ToUInt32(buffer.Sha256(), 0);
-            }
+            ms.Seek(0, SeekOrigin.Begin);
```
Comment (erikzhang) [inline src/neo/SmartContract/NefFile.cs:122]: I think we should still read the header.

Comment (shargon) [reply src/neo/SmartContract/NefFile.cs]: Fixed

Comment (shargon) [reply src/neo/SmartContract/NefFile.cs:127]: I think that the checksum should include the compiler version, what do you think?

File: src/neo/SmartContract/NefFile.cs [hunk 2]
```diff
@@ -110,20 +111,19 @@ public void Deserialize(BinaryReader reader)
...
-                ms.Read(buffer);
+            // Read header without CRC
 
-                return BitConverter.ToUInt32(buffer.Sha256(), 0);
-            }
+            Span<byte> buffer = stackalloc byte[HeaderSize - sizeof(uint)];
+            ms.Seek(0, SeekOrigin.Begin);
+            ms.Read(buffer);
+
+            return BitConverter.ToUInt32(buffer.Sha256().Sha256(), 0);
```
Comment (vncoelho) [inline src/neo/SmartContract/NefFile.cs:126]: `Hash256` is accessible here? https://github.com/neo-project/neo/blob/1af936929093721ec274167829761ddba825c86c/src/neo/Cryptography/Crypto.cs#L14-L17

File: src/neo/SmartContract/NefFile.cs [hunk 3]
```diff
@@ -14,7 +15,7 @@ namespace Neo.SmartContract
     /// | Version    | 16 bytes  | Compiler version (Mayor, Minor, Build, Version)            |
     /// | ScriptHash | 20 bytes  | ScriptHash for the script                                  |
     /// +------------+-----------+------------------------------------------------------------+
-    /// | Checksum   | 4 bytes   | Sha256 of the header (CRC)                                 |
+    /// | Checksum   | 4 bytes   | First four bytes of double SHA256 hash                     |
```
Comment (vncoelho) [inline src/neo/SmartContract/NefFile.cs:18]: "First four bytes of double SHA256 hash" -> "Hash256 if the first four bytes"

File: src/neo/SmartContract/NefFile.cs (hunk 2 above)
Comment (shargon) [reply src/neo/SmartContract/NefFile.cs:126]: Good! and done :)

Comment (Tommo-L) [reply src/neo/SmartContract/NefFile.cs:126]: ? it's still `buffer.Sha256().Sha256()`.

File: src/neo/SmartContract/NefFile.cs [hunk 4]
```diff
@@ -110,20 +111,19 @@ public void Deserialize(BinaryReader reader)
         /// <returns>Return checksum</returns>
         public static uint ComputeChecksum(NefFile file)
         {
-            using (var ms = new MemoryStream())
-            using (var wr = new BinaryWriter(ms))
-            {
-                file.Serialize(wr);
-                wr.Flush();
+            using var ms = new MemoryStream();
+            using var wr = new BinaryWriter(ms, Encoding.UTF8, false);
```
Comment (erikzhang) [inline src/neo/SmartContract/NefFile.cs:115]:
```suggestion
            using var wr = new BinaryWriter(ms, Utility.StrictUTF8, false);
```

**Pull request merged**